zeppelin-web/src/app/interpreter/interpreter.controller.js

```diff
@@ -22,6 +22,171 @@ angular.module('zeppelinWebApp').controller('InterpreterCtrl',
     $scope.showRepositoryInfo = false;
     $scope._ = _;
 
+    // auto complete related
+    $scope.suggestions = [];
+    $scope.selectIndex = -1;
+    var selectedUser = '';
+    var selectedUserIndex = 0;
+    var previousSelectedList = [];
+    var searchText = [];
+
+    function updatePreviousList() {
+      for (var i = 0; i < searchText.length; i++) {
+        previousSelectedList[i] = searchText[i];
+      }
+    }
+
+    function convertToString(setting) {
+      setting.option.users = searchText.join();
+    }
+
+    var checkIfSelected = function() {
+      if (($scope.suggestions.length === 0) &&
+        ($scope.selectIndex < 0 || $scope.selectIndex >= $scope.suggestions.length) ||
+        ($scope.suggestions.length !== 0 && ($scope.selectIndex < 0 || $scope.selectIndex >= $scope.suggestions.length))
```

Comment:
Long condition inside the `if` becomes quite hard to understand and follow here.

What do you think, could this be changed a bit to make it more readable i.e:

```javascript
if (suggestionSelected($scope.suggestions.length, $scope.selectIndex)) {
 //....
}

function suggestionSelected(len, i) {
  return (len === 0) && (i < 0 || i >= len) || 
         (len !== 0  && (i < 0 || i >= len))
}
```

And now it's easy to see how `suggestionSelected` can be further improved, keeping invariants.

Normalizing the parantheses according to [logical operator precedence in javascript](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Operators/Operator_Precedence) we get

```javascript
return (len === 0 && (i < 0 || i >= len)) || 
       (len !== 0 && (i < 0 || i >= len))
}
```

which, in turn, is the same as

```javascript
return (i < 0 || i >= len) && (len === 0 || len !== 0)
```

and as the right side of the `&&` is always true for an integer, we get

```javascript
function suggestionSelected(len, i) {
 return i < 0 || i >= len
}
```

At the end, this might not worth extracting to a separate function, but this shows the benefits we can get from it.
The result is much simpler to understand and reason about.

@astroshim what do you think?